R/plotAbundanceDensity.R

```diff
@@ -0,0 +1,171 @@
+                                                       n = n,
+                                                       colour_by = colour_by)
+        # Extracts the density data and aesthetic from the list
+        density_data <- density_data_list$density_data
+        colour_by <- density_data_list$colour_by
+        
+        # Gets the plot from plotter
+        plot_out <- .density_plotter(density_data = density_data, 
+                                     xlab = abund_values,
+                                     ylab = "Taxa",
+                                     colour_by = colour_by,
+                                     ...)
+        return(plot_out)
+    }
+)
+
+################################ HELP FUNCTIONS ################################
+
+.incorporate_density_data <- function(object, abund_values, n, colour_by){
+    # Gets the assay
+    mat <- assay(object, abund_values)
+    # Gets the most abundant taxa
+    top_taxa <- getTopTaxa(object, n)
+    # Subsets abundance table  by taking taxa of highest abundance
+    mat <- mat[top_taxa, , drop=FALSE]
+    # melt the data
+    density_data <- t(mat) %>%
+        as.data.frame() %>%
+        rownames_to_column("Sample") 
+    # Gets coloring information if 'colour_by' is not NULL
+    if (!is.null(colour_by)) {
+        # Gets information from colData
+        colour_out <- retrieveCellInfo(object, colour_by)
+        # Mirrors back the variable name, if a partial match was used
+        colour_by <- colour_out$name
+        # Adds information to the data frame that includes all density data
+        density_data$colour_by <- colour_out$value
+    }
+    cols <- c("Sample","colour_by")[c("Sample","colour_by") %in% colnames(density_data)]
+    density_data <- density_data %>%
+        pivot_longer(cols = !cols, names_to = "Y", values_to = "X")
+    # Converts taxa to factor. Order of levels is the opposite than in 'top_taxa'
+    # so that taxa with highest abundance is on top
+    density_data$Y <- factor( density_data$Y, rev(top_taxa) )
+    return(list(density_data = density_data, 
+                colour_by = colour_by))
+}
+
+.density_plotter <- function(density_data, 
+                             xlab = NULL,
+                             ylab = NULL,
+                             colour_by = NULL,
+                             layout = "point",
+                             point_alpha = 0.6,
+                             point_shape = 124,
+                             point_size = 2,
+                             text_size = 8,
+                             add_legend = TRUE,
+                             flipped = FALSE,
+                             add_x_text = FALSE){
+    # start plotting
+    plot_out <- ggplot(density_data, aes_string(x="X", y="Y")) +
+        xlab(xlab) +
+        ylab(ylab)
+    
+    # Layout "density" or "point", "density" will be added
+    if (layout == "point"){
```

Comment:
If I set `layout = "whatever"` and empty plot will be returned. There must be hard fallback, probably as an `else`

Reply:
In addition, the layout is `point` by default, but it does not really return a point shape by default. This is misleading. Does anything speak against using a jitter type of plot as the default and switch to density as an alternative?

Reply:
I think it is OK to put density as the default if this is easier.